astropy/units/format/ogip.py

```diff
@@ -205,34 +205,31 @@ def p_unit_expression(p):
                             | UNIT OPEN_PAREN complete_expression CLOSE_PAREN power numeric_power
                             | OPEN_PAREN complete_expression CLOSE_PAREN power numeric_power
             """
-            # If we run p[1] in cls._functions, it will try and parse each
-            # item in the list into a unit, which is slow. Since we know that
-            # all the items in the list are strings, we can simply convert
-            # p[1] to a string instead.
-            p1_str = str(p[1])
-
-            if p1_str in cls._functions and p1_str != "sqrt":
-                raise ValueError(
-                    f"The function '{p[1]}' is valid in OGIP, but not understood "
-                    "by astropy.units."
-                )
+            bad_function_message = (
+                "The function '{}' is valid in OGIP, but not understood "
+                "by astropy.units."
+            )
             bad_multiplication_message = (
                 "if '{0}{1}' was meant to be a multiplication, "
                 "it should have been written as '{0} {1}'."
             )
 
             if len(p) == 7:
-                if p1_str == "sqrt":
+                if p[1] == "sqrt":
                     p[0] = p[3] ** (0.5 * p[6])
+                elif p[1] in cls._functions:
+                    raise ValueError(bad_function_message.format(p[1]))
```

Comment:
I should add a test for this code branch, and I'll have to rebase anyways to incorporate whatever the fixes will be for the current problems on Windows.